### DISCUSSION TITLE
React Native 端，如何将未使用的原生模块从项目中移除的问题

### DISCUSSION THREAD

**Discussion by robinv8** [NervJS/taro#13592] [Q & A]
Taro 框架会内置一些 `expo` 的原生类库，这会导致在应用中包含一些无用原生模块，例如：应用中没有相机、录音的功能，但却需要其相关权限。

**Comment by robinv8** [comment]
其中 `autolinking` 可以通过如下方式解决，但在执行 `npm run android` 后 js bundle 中任然会包含原生类库
```json
{
  "expo": {
      "autolinking": {
        "android": {
          "exclude": [
            "expo-camera",
            "expo-image-picker",
            "expo-barcode-scanner",
            "expo-file-system"
          ]
        }
      }
   }
}
```

**Comment by zhiqingchen** [comment]
1. 需要在使用 taro 的时候，不使用的不支持的api。

https://github.com/NervJS/taro/blob/next/packages/babel-preset-taro/rn/index.js#L142-L163
接口和组件在使用的时候是转到对应的文件的。这样就能保证没使用的原生模块不被引入。当用到不支持的api或组件时，这个分析就会降级成全量引入。

2. 全量引入 Taro 的场景也不支持。如：

```js
import Taro from '@tarojs/taro';

…………

export default Taro
```

3. 不能用 Taro['xxx' + 'xxx']，类似这样的，无法被静态分析的，会导致降级到全量引用
